maro/simulator/scenarios/data_center/business_engine.py

```diff
@@ -0,0 +1,320 @@
+        self._event_buffer.register_event_handler(event_type=MaroEvents.TAKE_ACTION, handler=self._on_action_received)
+
+    def _update_vm_util(self):
+        """Update all live VMs CPU utilization.
+
+        The length of VMs utilization series could be difference among all VMs,
+        because index 0 represents the VM's CPU utilization at the tick it starts.
+        """
+
+        for vm in self._live_vm.values():
+            vm.util_cpu = vm.get_util(cur_tick=self._tick)
+
+    def _update_pm_util(self):
+        """Update CPU utilization occupied by total VMs on each PM."""
+
+        for pm in self._machines:
+            total_util_cpu: int = 0
+            for vm_id in pm.vm_set:
+                vm = self._live_vm[vm_id]
+                total_util_cpu += vm.util_cpu * vm.req_cpu / 100
+            util_cpu = total_util_cpu / pm.cap_cpu * 100
+            pm.update_util(tick=self._tick, util_cpu=util_cpu)
+            self._energy_consumption += self._cpu_util_to_energy_consumption(pm.util_cpu)
+
+    def _cpu_util_to_energy_consumption(self, cpu_util: float) -> float:
+        """Convert the CPU utilization to energy consumption."""
+
+        power: float = self._conf["calibration parameter"]
+        # NOTE: Energy comsumption parameters should refer to more research.
+        busy_power = self._conf["busy_power"]
+        idle_power = self._conf["idle_power"]
+        return idle_power + (busy_power - idle_power) * (2 * cpu_util - pow(cpu_util, power))
+
+    def _postpone_vm_requirement(self, postpone_type: PostponeType, vm_id: int, remaining_buffer_time: int):
+        """Postpone VM requirement."""
+
+        if remaining_buffer_time >= self._delay_duration:
+            if postpone_type == PostponeType.Resource:
+                self._total_latency.latency_due_to_resource += self._delay_duration
+            elif postpone_type == PostponeType.Agent:
+                self._total_latency.latency_due_to_agent += self._delay_duration
+
+            postpone_payload = self._pending_vm_req_payload[vm_id]
+            postpone_payload.remaining_buffer_time -= self._delay_duration
+            postpone_event = self._event_buffer.gen_cascade_event(
+                tick=self._tick + self._delay_duration,
+                event_type=Events.REQUIREMENTS,
+                payload=postpone_payload
+            )
+            self._event_buffer.insert_event(event=postpone_event)
+        else:
+            # Fail
+            # Pop out VM requirement payload.
+            self._pending_vm_req_payload.pop(vm_id)
+            # Add failed requirements.
+            self._failed_requirements += 1
+
+    def _get_valid_pm(self, vm_req_cpu: int) -> List[ValidPhysicalMachine]:
+        """Check all valid PMs.
+
+        There are three situations:
+            1. All PMs are empty.
+            2. Some PMs are empty but others are not.
+            3. All PMs are not empty.
+        Situation 1: Return the first empty PM info.
```

Comment:
why not always return the valid pms?

Reply:
After the discussion with @Jinyu-W, we thought that return "one empty" pm and "all other PMs with enough resource" could simplify agent's decision. Under this situation, agent only need to consider a PM (empty), and all the other PMs(not empty).

Reply:
I understand, the one empty pm may have two different meanings, one for all empty pms, another may be only empty pm is available.  I prefer always to return valid pms.